pandas/core/internals/blocks.py

```diff
@@ -423,7 +425,15 @@ def coerce_to_target_dtype(self, other) -> Block:
         and will receive the same block
         """
         new_dtype = find_result_type(self.values, other)
-
+        if warn_on_upcast and self.dtype != new_dtype:
```

Comment:
if a user ever gets here with `new_dtype=self.dtype` then something has gone wrong and they are going to get a RecursionError.  we should probably add a check for that and raise with a "please open a bug report"